Parsing the keys from config.yaml fails

Steps to reproduce:

Execute
```
./bin/alm --config config.yaml
```

The application crashes on startup with the following error:

```
{"err":"Invalid Key: Key must be PEM encoded PKCS1 or PKCS8 private key","level":"panic","msg":"failed to parse public token","pid":64594,"time":"2017-02-27 19:23:08"}
panic: (*logrus.Entry) (0x7e5660,0xc420697b30)

goroutine 1 [running]:
panic(0x7e5660, 0xc420697b30)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/almighty/almighty-core/vendor/github.com/Sirupsen/logrus.Entry.log(0xc420229cc0, 0xc4206b8000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc42061d800, ...)
	/Users/shoubhikbose/dev/alm-core-sbose/src/github.com/almighty/almighty-core/vendor/github.com/Sirupsen/logrus/entry.go:124 +0x423
github.com/almighty/almighty-core/vendor/github.com/Sirupsen/logrus.(*Entry).Panic(0xc420697ae0, 0xc42061d8f0, 0x1, 0x1)
	/Users/shoubhikbose/dev/alm-core-sbose/src/github.com/almighty/almighty-core/vendor/github.com/Sirupsen/logrus/entry.go:169 +0x124
github.com/almighty/almighty-core/vendor/github.com/Sirupsen/logrus.(*Entry).Panicln(0xc420697ae0, 0xc42061d988, 0x1, 0x1)
	/Users/shoubhikbose/dev/alm-core-sbose/src/github.com/almighty/almighty-core/vendor/github.com/Sirupsen/logrus/entry.go:264 +0xe9
github.com/almighty/almighty-core/log.Panic(0x0, 0x0, 0xc4206a9f20, 0x826b90, 0x1c, 0x0, 0x0, 0x0)
	/Users/shoubhikbose/dev/alm-core-sbose/src/github.com/almighty/almighty-core/log/log.go:202 +0x279
main.main()
	/Users/shoubhikbose/dev/alm-core-sbose/src/github.com/almighty/almighty-core/main.go:178 +0x234a
```

Added a test which showed the error - 

```
--- FAIL: TestGetTokenPrivateKeyFromConfigFile (0.00s)
        Error Trace:    configuration_blackbox_test.go:170
	Error:  	Expected nil, but got: &errors.errorString{s:"Invalid Key: Key must be PEM encoded PKCS1 or PKCS8 private key"}
```

--- a/config.yaml
+++ b/config.yaml
@@ -35,7 +35,7 @@ populate.commontypes: true
 # Authentication configuration
 # ----------------------------
 
-token.privatekey : >
+token.privatekey : |
                     -----BEGIN RSA PRIVATE KEY-----
                     MIIEpQIBAAKCAQEAnwrjH5iTSErw9xUptp6QSFoUfpHUXZ+PaslYSUrpLjw1q27O
                     DSFwmhV4+dAaTMO5chFv/kM36H3ZOyA146nwxBobS723okFaIkshRrf6qgtD6coT
@@ -64,7 +64,7 @@ token.privatekey : >
                     OCCAgsB8g8yTB4qntAYyfofEoDiseKrngQT5DSdxd51A/jw7B8WyBK8=
                     -----END RSA PRIVATE KEY-----
 
-token.publickey : >
+token.publickey : |
                     -----BEGIN PUBLIC KEY-----
                     MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAiRd6pdNjiwQFH2xmNugn
                     TkVhkF+TdJw19Kpj3nRtsoUe4/6gIureVi7FWqcb+2t/E0dv8rAAs6vl+d7roz3R

--- a/configuration/configuration_blackbox_test.go
+++ b/configuration/configuration_blackbox_test.go
@@ -3,14 +3,23 @@ package configuration_test
 import (
 	"fmt"
 	"os"
+	"strings"
 	"testing"
 
 	"net/http"
 
 	"github.com/almighty/almighty-core/configuration"
 	"github.com/almighty/almighty-core/resource"
+	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/goadesign/goa"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	varTokenPublicKey  = "token.publickey"
+	varTokenPrivateKey = "token.privatekey"
+	defaultConfigFile  = "config.yaml"
 )
 
 var reqLong *goa.RequestData
@@ -31,7 +40,7 @@ func TestMain(m *testing.M) {
 
 func resetConfiguration() {
 	var err error
-	config, err = configuration.GetConfigurationData()
+	config, err = configuration.NewConfigurationData("../config.yaml")
 	if err != nil {
 		panic(fmt.Errorf("Failed to setup the configuration: %s", err.Error()))
 	}
@@ -140,4 +149,50 @@ func TestGetKeycloakEndpointUserInfoSetByEnvVaribaleOK(t *testing.T) {
 	url, err = config.GetKeycloakEndpointUserInfo(reqShort)
 	assert.Nil(t, err)
 	assert.Equal(t, "userinfoEndpoint", url)
+}
+
+func TestGetTokenPrivateKeyFromConfigFile(t *testing.T) {
+	t.Parallel()
+
+	envKey := generateEnvKey(varTokenPrivateKey)
+
+	realEnvValue := os.Getenv(envKey) // could be "" as well.
+
+	os.Unsetenv(envKey)
+	defer os.Setenv(envKey, realEnvValue) // set it back before we leave.
+
+	// env variable NOT set, so we check with config.yaml's value
+
+	viperValue := config.GetTokenPrivateKey()
+	assert.NotNil(t, viperValue)
+
+	parsedKey, err := jwt.ParseRSAPrivateKeyFromPEM(viperValue)
+	require.Nil(t, err)
+	assert.NotNil(t, parsedKey)
+}
+
+func TestGetTokenPublicKeyFromConfigFile(t *testing.T) {
+	t.Parallel()
+
+	envKey := generateEnvKey(varTokenPublicKey)
+	realEnvValue := os.Getenv(envKey) // could be "" as well.
+
+	os.Unsetenv(envKey)
+	defer func() {
+		os.Setenv(envKey, realEnvValue)
+		resetConfiguration()
+	}()
+
+	// env variable is now unset for sure, this will lead to the test looking up for
+	// value in config.yaml
+	viperValue := config.GetTokenPublicKey()
+	assert.NotNil(t, viperValue)
+
+	parsedKey, err := jwt.ParseRSAPublicKeyFromPEM(viperValue)
+	require.Nil(t, err)
+	assert.NotNil(t, parsedKey)
+}
+
+func generateEnvKey(yamlKey string) string {
+	return "ALMIGHTY_" + strings.ToUpper(strings.Replace(yamlKey, ".", "_", -1))
 }

--- a/configuration/configuration_blackbox_test.go
+++ b/configuration/configuration_blackbox_test.go
@@ -152,14 +152,15 @@ func TestGetKeycloakEndpointUserInfoSetByEnvVaribaleOK(t *testing.T) {
 }
 
 func TestGetTokenPrivateKeyFromConfigFile(t *testing.T) {
-	t.Parallel()
 
 	envKey := generateEnvKey(varTokenPrivateKey)
-
 	realEnvValue := os.Getenv(envKey) // could be "" as well.
 
 	os.Unsetenv(envKey)
-	defer os.Setenv(envKey, realEnvValue) // set it back before we leave.
+	defer func() {
+		os.Setenv(envKey, realEnvValue)
+		resetConfiguration()
+	}()
 
 	// env variable NOT set, so we check with config.yaml's value
 
@@ -172,7 +173,6 @@ func TestGetTokenPrivateKeyFromConfigFile(t *testing.T) {
 }
 
 func TestGetTokenPublicKeyFromConfigFile(t *testing.T) {
-	t.Parallel()
 
 	envKey := generateEnvKey(varTokenPublicKey)
 	realEnvValue := os.Getenv(envKey) // could be "" as well.

--- a/configuration/configuration_blackbox_test.go
+++ b/configuration/configuration_blackbox_test.go
@@ -17,9 +17,10 @@ import (
 )
 
 const (
-	varTokenPublicKey  = "token.publickey"
-	varTokenPrivateKey = "token.privatekey"
-	defaultConfigFile  = "config.yaml"
+	varTokenPublicKey           = "token.publickey"
+	varTokenPrivateKey          = "token.privatekey"
+	defaultConfigFilePath       = "../config.yaml"
+	defaultValuesConfigFilePath = "" // when the code defaults are to be used, the path to config file is ""
 )
 
 var reqLong *goa.RequestData
@@ -27,7 +28,7 @@ var reqShort *goa.RequestData
 var config *configuration.ConfigurationData
 
 func TestMain(m *testing.M) {
-	resetConfiguration()
+	resetConfiguration(defaultConfigFilePath)
 
 	reqLong = &goa.RequestData{
 		Request: &http.Request{Host: "api.service.domain.org"},
@@ -38,9 +39,11 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
 
-func resetConfiguration() {
+func resetConfiguration(configPath string) {
 	var err error
-	config, err = configuration.NewConfigurationData("../config.yaml")
+
+	// calling NewConfigurationData("") is same as GetConfigurationData()
+	config, err = configuration.NewConfigurationData(configPath)
 	if err != nil {
 		panic(fmt.Errorf("Failed to setup the configuration: %s", err.Error()))
 	}
@@ -66,11 +69,11 @@ func TestGetKeycloakEndpointAuthSetByEnvVaribaleOK(t *testing.T) {
 	env := os.Getenv("ALMIGHTY_KEYCLOAK_ENDPOINT_AUTH")
 	defer func() {
 		os.Setenv("ALMIGHTY_KEYCLOAK_ENDPOINT_AUTH", env)
-		resetConfiguration()
+		resetConfiguration(defaultValuesConfigFilePath)
 	}()
 
 	os.Setenv("ALMIGHTY_KEYCLOAK_ENDPOINT_AUTH", "authEndpoint")
-	resetConfiguration()
+	resetConfiguration(defaultValuesConfigFilePath)
 
 	url, err := config.GetKeycloakEndpointAuth(reqLong)
 	assert.Nil(t, err)
@@ -101,11 +104,11 @@ func TestGetKeycloakEndpointTokenSetByEnvVaribaleOK(t *testing.T) {
 	env := os.Getenv("ALMIGHTY_KEYCLOAK_ENDPOINT_TOKEN")
 	defer func() {
 		os.Setenv("ALMIGHTY_KEYCLOAK_ENDPOINT_TOKEN", env)
-		resetConfiguration()
+		resetConfiguration(defaultValuesConfigFilePath)
 	}()
 
 	os.Setenv("ALMIGHTY_KEYCLOAK_ENDPOINT_TOKEN", "tokenEndpoint")
-	resetConfiguration()
+	resetConfiguration(defaultValuesConfigFilePath)
 
 	url, err := config.GetKeycloakEndpointToken(reqLong)
 	assert.Nil(t, err)
@@ -136,11 +139,11 @@ func TestGetKeycloakEndpointUserInfoSetByEnvVaribaleOK(t *testing.T) {
 	env := os.Getenv("ALMIGHTY_KEYCLOAK_ENDPOINT_USERINFO")
 	defer func() {
 		os.Setenv("ALMIGHTY_KEYCLOAK_ENDPOINT_USERINFO", env)
-		resetConfiguration()
+		resetConfiguration(defaultValuesConfigFilePath)
 	}()
 
 	os.Setenv("ALMIGHTY_KEYCLOAK_ENDPOINT_USERINFO", "userinfoEndpoint")
-	resetConfiguration()
+	resetConfiguration(defaultValuesConfigFilePath)
 
 	url, err := config.GetKeycloakEndpointUserInfo(reqLong)
 	assert.Nil(t, err)
@@ -159,9 +162,10 @@ func TestGetTokenPrivateKeyFromConfigFile(t *testing.T) {
 	os.Unsetenv(envKey)
 	defer func() {
 		os.Setenv(envKey, realEnvValue)
-		resetConfiguration()
+		resetConfiguration(defaultValuesConfigFilePath)
 	}()
 
+	resetConfiguration(defaultConfigFilePath)
 	// env variable NOT set, so we check with config.yaml's value
 
 	viperValue := config.GetTokenPrivateKey()
@@ -180,8 +184,10 @@ func TestGetTokenPublicKeyFromConfigFile(t *testing.T) {
 	os.Unsetenv(envKey)
 	defer func() {
 		os.Setenv(envKey, realEnvValue)
-		resetConfiguration()
+		resetConfiguration(defaultValuesConfigFilePath)
 	}()
+
+	resetConfiguration(defaultConfigFilePath)
 
 	// env variable is now unset for sure, this will lead to the test looking up for
 	// value in config.yaml

